### ISSUE TITLE
`${path}` doesn't escape `$` signs for PowerShell's "Show current file in File Explorer" command

### ISSUE THREAD

**Issue by Sneakpeakcss** [mpvnet-player/mpv.net#580]
**Describe the bug**

The command to open the file in explorer:
``` powershell

run powershell -command "explorer.exe '/select,' ( \"${path}\" -replace '/', '\\' )"

```
breaks the moment there's `$` sign in the path. I've tried to [replace it ](https://stackoverflow.com/a/34489647), but that didn't work for me. It seems like the moment `${path}` is replaced, it already cut everything after `$`.

**To Reproduce**

Opening:
C:\test\test test\\**tes.mp4**               :heavy_check_mark:
C:\test\test test\\**test123.mp4**      :heavy_check_mark:
C:\test\test test\\**tes$ t123.mp4**  :heavy_check_mark: (space after `$`)
C:\test\test test\\**tes**$t123<strong>.mp4</strong>  :x: (opens `This PC` directory, or a 'matching' file if it exists, in this case  `C:\test\test test\tes.mp4`  )

1. Play a video with a `$` sign in the path and press `e` 

**Expected behavior**
`$` should probably be escaped (`` `$ ``) before passing the command to PowerShell.

**Additional context**
1. mpv.net v7
2. Windows 22H2

**Comment by stax76** [comment]
Do you often have such files?

**Comment by Sneakpeakcss** [comment]
Not that often, but the files that i work with have predefined names, so from time to time a dollar sign shows up in them. It also causes problems in folder names, so it's not only exclusive to files. I'll check if i can find the mentioned Lua script later, though it would be good to know its name or original url to it.

**Comment by Sneakpeakcss** [comment]
Seems like it was completely lost, few sites managed to [cache](https://cc.bingj.com/cache.aspx?q=url:https://github.com/ayghub/open-dir&d=4740226878942857&w=uXvhru4NLpBrmzdlg-B63d5oIdhXrs4H) the repo, but not the script itself.

It was simpler to remake it than digging through obscure parts of internet:

`openInExplorer.lua`
```
e           script-message-to openInExplorer open-in-explorer  #menu: Tools > Show current file in File Explorer
```
<sub>PowerShell:</sub>
``` lua
local function open_in_explorer()
    local path = mp.get_property("path")
    if path ~= nil then
        local ps_code = [[
            explorer.exe /select,"__path__"
        ]]
        local path = string.gsub(path, "[%(%)%.%+%-%*%?%[%]%^%$%%]", "%%%1")
        local ps_path = string.gsub(path, "([$`])", "`%1")      -- escape dollar/backtick signs for PowerShell
        local ps_code = string.gsub(ps_code, "__path__", ps_path)
        mp.command_native({
            name = "subprocess",
            playback_only = false,
            args = { 'powershell', '-NoProfile', '-Command', ps_code },
        })
    end
end

mp.add_key_binding(nil, "open-in-explorer", open_in_explorer)
```

However, at this point it's probably better to stop using PowerShell for this:

<sub>direct:</sub>
```lua
local function open_in_explorer()
    local path = mp.get_property("path")
    if path ~= nil then
        mp.command_native({
            name = "subprocess",
            playback_only = false,
            args = { 'explorer', '/select,',  path },
        })
    end
end

mp.add_key_binding(nil, "open-in-explorer", open_in_explorer)
```

Both manage to open ```( ) % . + - [ ^ $ ' ` .mp4``` and  ```tt(tt)tt%tt.tt+tt-tt[tt^tt$tt'tt`tt .mp4``` without any issues, though the second one is slightly faster:

 
![test](https://github.com/mpvnet-player/mpv.net/assets/77424331/fec613d0-88f6-4ec5-af9e-a0ff7d9a7487)
so i'm not sure if anyone would need the PowerShell version at that point.
___

Seems like this issue might be closed then, unless you want to keep it open for one reason or another.

**Comment by stax76** [comment]
PowerShell is great, but starts very slow, compared to cmd, bash, nushell, Lua, Python etc.

I removed the PowerShell one-liner from the defaults because of the $ character limitation/bug and because there aren't any Lua scripts included with mpv.net. It should be fine having only a minimal default feature set, people who need extra features can install a script. Including Lua scripts is a possibility in the future, but there are more important tasks, like Internationalization.

With some modification, I added the code to my private.lua script.

```Lua
-- e  script-binding  fooscript/open-in-explorer
mp.add_key_binding(nil, "open-in-explorer", function ()
    local path = mp.get_property("path")
    if path == nil or string.find(path, "://") then return end
    path = string.gsub(path, "/", "\\")
    mp.command_native({
        name = "subprocess",
        playback_only = false,
        args = { 'explorer', '/select,',  path },
    })
end)
```

If you don't have a mpv-scripts repo, you can create a GitHub Gist. And add it to the mpv wiki. I think it's very useful, I use it frequently for music files. Thanks for working on a solution.

**Comment by Sneakpeakcss** [comment]
That’s a nifty way to include the keybind. I was so focused on the main `$` issue that i didn’t think much about the scripts structure, or even why bother to do it in PS at all, until it was already done.

However, even at the beginning i noticed and thought about the forward slash replacemet, but isn't this technically redundant, since this script is exclusive to Windows? 

I remember googling a bit and finding some general references to "network paths", but i don't use any of that, so i wasn't able to check how mpv handles them. That one made me drop my initial idea of using `path:match("^%a:\\")` just to check for volume letter and leave it at that.

Are there any situations on Windows where mpv provides path with `/` ? Because if so, i would rather fix my local version before even trying to add it to wiki (i already managed to shoot myself in the foot for not managing protocol paths):


``` lua
mp.add_key_binding(nil, "open-in-explorer", function()
    local path = mp.get_property("path")
    if path ~= nil and not path:match("^%a[%a%d-_]+://") then
        mp.command_native({
            name = "subprocess",
            playback_only = false,
            args = { 'explorer', '/select,',  path },
        })
    else
        mp.osd_message("Invalid path: " .. (path or "No path provided"))
    end
end)
```

_Bonus feedback message to avoid mashing the key 20 times in a row, just to be sure…_

 
When it comes to playing video files, i barely managed to properly transition from `mpc-hc` to `mpv`, and i didn't drop that idea completely mostly because of `mpv.net` and `uosc` (<sub> not to mention that Lua doesn’t require compiling, so it was stupidly easy for someone who doesn’t program at all to butcher other people’s code when required</sub>), so at this point it's hard for me to even imagine transitioning from `foobar2000` with `JScript Panel 3`/`ColumnsUI` for music files. Mpv seems very "resistant" when it comes to things like proper playlist managment.

**Comment by stax76** [comment]
> However, even at the beginning i noticed and thought about the forward slash replacement, but isn't this technically redundant, since this script is exclusive to Windows?

In this case particular case, backslashes are necessary, you can test it with the following PS command:

```PowerShell
explorer '/select,' C:\Windows/explorer.exe # this does not work

explorer '/select,' C:\Windows\explorer.exe # this works
```

It was mentioned in a discussion I remembered, probably [here](https://github.com/mpv-player/mpv/issues/6560#issuecomment-473764245).

Either way, normalizing is probably better, it's unclear if Explorer and Win32 stores the path normalized, probably it does, but if it doesn't, it could cause issues, for instance with not commonly used explorer features or shell extensions.

> Are there any situations on Windows where mpv provides path with / ?

Yes, mpv doesn't normalize in many cases. Maybe add a comment, 'Explorer requires normalization in this case', so it's clear for everybody.

For music, I used MediaMonkey and MusicBee before migrating to mpv/mpv.net. There are music playlists on YouTube with all the music I liked in my lifetime, almost 1000 songs.

https://www.youtube.com/@stax76/playlists

**Comment by Sneakpeakcss** [comment]
> For music, I used MediaMonkey and MusicBee before migrating to mpv/mpv.net. There are music playlists on YouTube with all the music I liked in my lifetime, almost 1000 songs.
> 
> https://www.youtube.com/@stax76/playlists

Nice, some of them reminded me of [Emancipator](https://www.youtube.com/watch?v=LXaeQw96Fx4) or [The Glitch Mob](https://www.youtube.com/watch?v=ezk_dD2Ia-w) when it comes to chill/electronic music.

> In this case particular case, backslashes are necessary, you can test it with the following PS commands:
> 
> ```powershell
> explorer '/select,' C:\Windows/explorer.exe # this does not work
> 
> explorer '/select,' C:\Windows\explorer.exe # this works
> ```

Yea, that was the main reason i started thinking about this. I wasn't able find examples to test it and see if mpv doesn't just replace them automatically when providing `path` on windows, and now i found [this](https://mpv.io/manual/stable/#protocols), so apparanely this: `file://C:/test/test.mp4` is a valid path in mpv.

![image](https://github.com/mpvnet-player/mpv.net/assets/77424331/51266cfa-cdc9-442b-bb62-5c4f72d1d7bb)

That's a good enough example that proves the point, so there’s no reason not to include it:

``` lua
mp.add_key_binding(nil, "open-in-explorer", function()
    local path = mp.get_property("path")
    if path ~= nil and not path:match("^%a[%a%d-_]+://") then
        path = string.gsub(path, "/", "\\")
        mp.command_native({
            name = "subprocess",
            playback_only = false,
            args = { 'explorer', '/select,',  path },
        })
    else
        mp.osd_message("Invalid path: " .. (path or "No path provided"))
    end
end)
```

Even without this, at least it didn't just error out like it happened when i tried using it on youtube link before handling those cases, so that's nice.

**Comment by Sneakpeakcss** [comment]
This didn't take long... [new-old](https://gist.github.com/Sneakpeakcss/05a97d509b8be67a6f11400b0bee54ab?permalink_comment_id=4765886#gistcomment-4765886) issues coming back in force, but this time even stronger. `$` isn't an issue anymore, but apparently a comma in a path that doesn't contain any spaces also had problems from the beginning:

> `C:\test\test2\aa,aa.mp4` this will open "This PC" directory
> `C:\test\test2\,aaaa.mp4` this will open "test" directory and select test2 folder
> `C:\test\test2\..,aaaa.mp4` uh huh… Amazing way of navigating directories
> 
> Everything starts working fine the moment there's a space anywhere in the path.

I don't know how mpv executes that command when the subprocess is "direct" like in the script above, but i can only assume it's similar if not the same as in cmd version, so if we believe ProcessExplorer this is what's being passed from mpv:

PowerShell:
<sub>`args = { 'powershell', '-NoProfile', '-Command', ps_code },`</sub>
``` powershell
"powershell" -NoProfile -Command "  explorer.exe /select,\"C:\test\test2\,aaaa.mp4\"  "  #❌ opens 'test' directory and selects 'test2' folder
"powershell" -NoProfile -Command "  explorer.exe /select,\"C:\test\test2\aa,aa.mp4\"  "  #❌ opens 'This PC' directory
"powershell" -NoProfile -Command "  explorer.exe /select,\"C:\test\test2\aa ,aa.mp4\"  " #✔️ space in path makes it work
```

cmd:
<sub>`args = { 'cmd', '/c', 'explorer', '/select,',  path },`</sub>
``` cmd
"cmd" /c explorer /select, C:\test\test2\,aaaa.mp4    ❌ opens 'test' directory and selects 'test2' folder
"cmd" /c explorer /select, C:\test\test2\aa,aa.mp4    ❌ opens 'This PC' directory
"cmd" /c explorer /select, "C:\test\test2\aa ,aa.mp4" ✔️ space in path makes it work
```

This works in all cases when done manually from command prompt:

``` cmd
explorer.exe /select,"C:\Users\omg\Desktop\test\test2\aa,aa.mp4"  ✔️ works in every single case
```

Well, that's promising. Since mpv doesn't pass any quotation marks in the problematic cases lets just do it in the script and see what mpv has to say about it:
<sub>`args = { 'cmd', '/c', 'explorer', '/select,',  '"' .. path .. '"' },`</sub>
``` cmd
"cmd" /c explorer /select, "\"C:\test\test2\aa,aa.mp4\""
```
Amazing… It doesn't work.
 

PowerShell version is even more mystical when it comes to getting `explorer.exe /select,` to work, even in the window manually:

```powershell
Start-Process -FilePath explorer.exe -ArgumentList "/select,`"C:\test\test2\aa,aa.mp4`""  #✔️ works in every single case
```
This is the only way i managed to get it to work, and that's from the PS window directly…

This seems to work:
``` lua
mp.add_key_binding(nil, "open-in-explorer", function()
    local path = mp.get_property("path")
    if path ~= nil and path:match("^%a:[/\\]") and not path:match("^%a[%a%d-_]+://") then
        local ps_code = [[
            Start-Process -FilePath explorer.exe -ArgumentList "/select,`"__path__`""
        ]]
        local path = string.gsub(path, "/", "\\")
        local path = string.gsub(path, "[%(%)%.%+%-%*%?%[%]%^%$%%]", "%%%1")
        local ps_path = string.gsub(path, "([$`])", "`%1")      -- escape dollar/backtick signs for PowerShell
        local ps_code = string.gsub(ps_code, "__path__", ps_path)
        mp.command_native({
            name = "subprocess",
            playback_only = false,
            args = { 'powershell', '-NoProfile', '-Command', ps_code },
        })
    else
        mp.osd_message("Invalid path: " .. (path or "No path provided"))
    end
end)
```

![beatingRecords](https://github.com/mpvnet-player/mpv.net/assets/77424331/a49956ac-37dc-48ed-bede-97f8b40935be)

We're breaking records… At this point this seems like a case of "just Windows / mpv things" and pick your poison. It would be far more preferable to get mpv to pass it with a single set of double quotes through cmd in all cases, but i don't think it's possible.

**Comment by stax76** [comment]
It's awesome that you could find a working solution, normally I can make my command lines work, but here I probably would have failed.

The main issue maybe is the crappy Explorer CLI.

For command line debugging, I sometimes use this tool: 
https://github.com/stax76/show-args

Another one that can be useful:
https://github.com/stax76/run-hidden

A utility I used and updated today:
https://github.com/stax76/command-line-to-exe

**Comment by Sneakpeakcss** [comment]
> For command line debugging, I sometimes use this tool: https://github.com/stax76/show-args

I found exactly the same repo, but of course only after i finished fighting with ProcessExplorer that wasn't always showing the launched cmd.exe/powershell.exe processes to grab the passed commands. I really hope that i won't have to ever use it, because at this point i'm having "proper character escaping while passing argument through different application" flashbacks.

> Another one that can be useful: https://github.com/stax76/run-hidden

Well, doesn't that look familiar... _I was [there](https://www.github.com/PowerShell/PowerShell/issues/3028#issuecomment-1205088096) in the before times_, while suffering like everyone else trying to hide the window. That's probably the starting point for my mentioned flashbacks when i discovered `Flow Launcher` and was trying to pass arguments to .bat/.ps1 scripts in `Plugin Runner` without any window flashing. That one and your `Favorites` are my most often used plugins.

___
This is the moment where i say that i feel like a fool:
>Everything starts working fine the moment there's a space anywhere in the path.

<sub>`args = { 'cmd', '/c', 'explorer', '/select,',  path .. ' ' },`</sub>
``` cmd
"cmd" /c explorer /select, "C:\test\test2\,aaaa.mp4 "   ✔️
"cmd" /c explorer /select, "C:\test\test2\aa,aa.mp4 "   ✔️
"cmd" /c explorer /select, "C:\test\test2\..,aaaa.mp4 " ✔️
```
 

```lua
-- open-in-explorer.lua

mp.add_key_binding(nil, "open-in-explorer", function()
    local path = mp.get_property("path")
    if path ~= nil and not path:match("^%a[%a%d-_]+://") then
        path = string.gsub(path, "/", "\\")
        mp.command_native({
            name = "subprocess",
            playback_only = false,
            args = { 'explorer', '/select,',  path .. ' ' },
        })
    else
        mp.osd_message("Invalid path: " .. (path or "No path provided"))
    end
end)
```

We're spared the pains of PowerShell again. I can't wait for another edge case that will make everything explode.
